azure-iot-device/tests/common/pipeline/test_pipeline_ops_base.py

```diff
@@ -7,64 +7,268 @@
+        assert op.request_type == init_kwargs["request_type"]
+
+    @pytest.mark.it("Initializes 'method' attribute with the provided 'method' parameter")
+    def test_method_type(self, cls_type, init_kwargs):
+        op = cls_type(**init_kwargs)
+        assert op.method == init_kwargs["method"]
+
+    @pytest.mark.it(
+        "Initializes 'resource_location' attribute with the provided 'resource_location' parameter"
+    )
+    def test_resource_location(self, cls_type, init_kwargs):
+        op = cls_type(**init_kwargs)
+        assert op.resource_location == init_kwargs["resource_location"]
+
+    @pytest.mark.it(
+        "Initializes 'request_body' attribute with the provided 'request_body' parameter"
+    )
+    def test_request_body(self, cls_type, init_kwargs):
+        op = cls_type(**init_kwargs)
+        assert op.request_body == init_kwargs["request_body"]
+
+    @pytest.mark.it("Initializes 'status_code' attribute to None")
+    def test_status_code(self, cls_type, init_kwargs):
+        op = cls_type(**init_kwargs)
+        assert op.status_code is None
+
+    @pytest.mark.it("Initializes 'response_body' attribute to None")
+    def test_response_body(self, cls_type, init_kwargs):
+        op = cls_type(**init_kwargs)
+        assert op.response_body is None
+
+
+pipeline_ops_test.add_operation_tests(
+    test_module=this_module,
+    op_class_under_test=pipeline_ops_base.RequestAndResponseOperation,
+    op_test_config_class=RequestAndResponseOperationTestConfig,
+    extended_op_instantiation_test_class=RequestAndResponseOperationInstantiationTests,
 )
-pipeline_data_object_test.add_operation_test(
-    cls=pipeline_ops_base.RequestOperation,
-    module=this_module,
-    positional_arguments=[
-        "request_type",
-        "method",
-        "resource_location",
-        "request_body",
-        "request_id",
-        "callback",
-    ],
+
+
+class RequestOperationTestConfig(object):
+    @pytest.fixture
+    def cls_type(self):
+        return pipeline_ops_base.RequestOperation
+
+    @pytest.fixture
+    def init_kwargs(self, mocker):
+        kwargs = {
+            "method": "SOME_METHOD",
+            "resource_location": "some/resource/location",
+            "request_type": "some_request_type",
+            "request_body": "some_request_body",
+            "request_id": "some_request_id",
+            "callback": mocker.MagicMock(),
+        }
+        return kwargs
+
+
+class RequestOperationInstantiationTests(RequestOperationTestConfig):
+    @pytest.mark.it("Initializes the 'method' attribute with the provided 'method' parameter")
+    def test_method(self, cls_type, init_kwargs):
+        op = cls_type(**init_kwargs)
+        assert op.method == init_kwargs["method"]
+
+    @pytest.mark.it(
+        "Initializes the 'resource_location' attribute with the provided 'resource_location' parameter"
+    )
+    def test_resource_location(self, cls_type, init_kwargs):
+        op = cls_type(**init_kwargs)
```

Comment:
>op = cls_type(**init_kwargs) [](start = 8, length = 28)

nit: this line is repeated many many times, op could be a fixture, either in this class, RequestOperationTestConfig, or in some base class.  If you do this, you can remove cls_type from the test parameters while you add op.   Feel free to ignore.  I tend to over-design.

Reply:
The rationale here is that as the test is testing the instantiation, the instantiation should be in the test

---
In reply to: [347139500](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r347139500) [](ancestors = 347139500)